### One Factor ANOVA

In [1]:
# One Factor ANOVA
def Anova_pvalue_HSD(sample_data,alpha):
    from scipy.stats import f
    import statistics
    from statsmodels.stats.libqsturng import qsturng
    import statistics,math,numpy
    
    joined_samples = numpy.concatenate(sample_data)
    EX_sq = [item**2 for item in joined_samples]
    total_EX_sq = sum(EX_sq)
    print(f'total_EX_sq = {total_EX_sq}')

    N = 0
    k = len(sample_data)
    for s in sample_data:
        N += len(s)
        n = len(s)
    print(f'N = {N}\nn = {n}\nk = {k}')
    
    df_between = k-1
    df_within = N-k
    print(f'df_between = {df_between}\ndf_within = {df_within}')

    squares = []
    G_raw = []
    SS_between_raw = []
    for s in sample_data:
        G_raw.append(sum(s))
        SS_between_raw.append((sum(s)**2)/n)

        for item in s:
            squares.append(item**2)
        
    G = round(sum(G_raw),3)
    print(f'G = {G}')

    SS_between = round(sum(SS_between_raw) - (G**2/N),3)
    print(f'SS_between = {SS_between}')

    sum_of_squares = sum(squares)

    SS_within = round(sum_of_squares - sum(SS_between_raw),3)
    print(f'SS_within = {SS_within}')

    SS_total = round(sum_of_squares - (G**2/N),3)
    print(f'SS_total = {SS_total}')

    accuracy_check = (SS_between + SS_within) == SS_total
    print(f'accuracy_check = {accuracy_check}')

    MS_within = round(SS_within/df_within,3)
    print(f'MS_within = {MS_within}')

    MS_between = round(SS_between/df_between,3)
    print(f'MS_between = {MS_between}')

    SS_total = SS_within + SS_between
    print(f'SS_total = {SS_total}')

    F_critical_value = round(f.ppf(1 - alpha, df_between, df_within),3)
    print(f'F_critical_value = {F_critical_value}')

    F_observed = round(MS_between/MS_within,3)
    print(f'F_observed = {F_observed}')

    if F_observed >= F_critical_value:
        print(f'F_observed = {F_observed} >= {F_critical_value}, therefore we will REJECT the Null Hypothesis')
    else:
        print(f'F_observed = {F_observed} < {F_critical_value}, therefore we will RETAIN the Null Hypothesis')

    p_value = round(1 - f.cdf(F_observed, df_between, df_within),7)

    if p_value < 0.05:
        print(f'p_value = {p_value} is LESS than {0.05}, therefore the F observed is SIGNIFICANT')
        proportion_of_explained_variance = round(SS_between/SS_total,3)
        print(f'proportion_of_explained_variance = {proportion_of_explained_variance}')
        q = round(qsturng(1-alpha, k, df_within),3)
        print(f'q = {q}')

        Tukeys_HSD_test = round(q*((MS_within/n)**0.5),3)
        print(f'Tukeys_HSD_test = {Tukeys_HSD_test}')

        means = [round(statistics.mean(data),3) for data in sample_data]
        print(f'means = {means}')

        max_mean = max(means)
        min_mean = min(means)
    
        Cohens_d = round((max_mean-min_mean)/MS_within**0.5,3)
        print(f'Cohens_d = {Cohens_d}')
    
    else:
        print(f'p_value = {p_value} is GREATER than {0.05}, therefore the F observed is INSIGNIFICANT')

#### Repeated Measures ANOVA

In [2]:
# Repeated Measures ANOVA
def Repeated_ANOVA(the_data,alpha):
    import statistics,math,numpy
    from scipy.stats import f
    import statsmodels.stats.libqsturng as qsturng

    joined_data = numpy.concatenate(the_data)
    print(f'joined_data = {joined_data}')

    k = len(the_data)
    print(f'k = {k}')

    N = len(joined_data)
    print(f'N = {N}')

    n = len(the_data[0])
    print(f'n = {n}')

    df_total = N-1
    print(f'df_total = {df_total}')

    df_between = k-1
    print(f'df_between = {df_between}')

    df_within = N-k
    print(f'df_within = {df_within}')

    df_subject = n-1
    print(f'df_subject = {df_subject}')

    df_error = df_within - df_subject
    print(f'df_error = {df_error}')
    
    T = [round(sum(data),3) for data in the_data]
    print(f'T = {T}')

    G = round(sum(T),3)
    print(f'G = {G}')

    T_sq_over_n = [round(t**2/n,3) for t in T]
    print(f'T_sq_over_n = {T_sq_over_n}')

    G_sq_over_N = round(G**2/N,3)
    print(f'G_sq_over_N = {G_sq_over_N}')

    SS_between = round(sum(T_sq_over_n) - G_sq_over_N,3)
    print(f'SS_between = {SS_between}')

    sum_X_sq = round(sum([num**2 for num in joined_data]),3)
    print(f'sum_X_sq = {sum_X_sq}')

    SS_within = round(sum_X_sq - sum(T_sq_over_n),3)
    print(f'SS_within = {SS_within}')

    # Subjects
    subjects = []
    n = 0
    while n < len(the_data[0]):
        temp = []
        subjects.append(temp)
        for i in range(len(the_data)):
            temp.append(the_data[i][n])
        n+=1
    print(f'subjects = {subjects}')

    subject_totals = [round(sum(list),3) for list in subjects]
    print(f'subject_totals = {subject_totals}')

    T_sq_subject = [round(num**2/k,3) for num in subject_totals]
    print(f'T_sq_subject = {T_sq_subject}')

    SS_subject = round(sum(T_sq_subject) - G_sq_over_N,3)
    print(f'SS_subject = {SS_subject}')

    SS_error = round(SS_within - SS_subject,3)
    print(f'SS_error = {SS_error}')

    SS_total = round(sum_X_sq - G_sq_over_N,3)
    print(f'SS_total = {SS_total}')

    MS_between = round(SS_between/df_between,3)
    print(f'MS_between = {MS_between}')

    MS_error = round(SS_error/df_error,3)
    print(f'MS_error = {MS_error}')

    F_ratio = round(MS_between/MS_error,3)
    print(f'F_ratio = {F_ratio}')

    F_critical_value = round(f.ppf(1 - alpha, df_between, df_error),3)
    print(f'F_critical_value = {F_critical_value}')

    if F_ratio >= F_critical_value:
        print(f'F_ratio = {F_ratio} is GREATER or EQUAL to F_critical_value = {F_critical_value}, therefore we will REJECT the Null Hypothesis')
    else:
        print(f'F_ratio = {F_ratio} is LESS than the F_critical_value = {F_critical_value}, therefore we will RETAIN the Null Hypothesis')

    eta_sq_p = round(SS_between/(SS_between + SS_error),3)
    print(f'proportion of explained variance = {eta_sq_p}')

    means = [round(statistics.mean(list),3) for list in the_data]
    print(f'means = {means}')

    q = round(qsturng.qsturng(1 - alpha, k, df_error),3)
    print(f'q = {q}')

    HSD = round(q*((MS_error/n)**0.5),3)
    print(f'HSD = {HSD}')
    
    means.sort(reverse=True)
    Cohens_ds = [round((max(means)-means[n])/MS_error**0.5,3) for n in range(1,len(means))]
    print(f"Cohen's d's = {Cohens_ds}")

### Two Factor ANOVA

In [29]:
def two_factor_ANOVA(the_data,alpha,r):
    import statistics,numpy,math
    import statsmodels.stats.libqsturng as qsturng
    from scipy.stats import f
    
    joined = numpy.concatenate(the_data)
    print(f'joined = {joined}')
    
    rn = len(the_data[0])
    print(f'rn = {rn}')

    N = len(joined)
    print(f'N = {N}')

    G = sum(joined)
    print(f'G = {G}')

    c = len(the_data)
    print(f'c = {c}')
    
    cn = len(the_data)*2
    print(f'cn = {cn}')
    
    X_sq = [round(num**2,3) for num in joined]
    print(f'X_sq = {X_sq}')

    sum_X_sq = sum(X_sq)
    print(f'sum_X_sq = {sum_X_sq}')

    G_sq_over_N = G**2/N
    print(f'G_sq_over_N = {G_sq_over_N}')
    
    SS_total = round(sum_X_sq - G_sq_over_N,3)
    print(f'SS_total = {SS_total}')
    
    column_totals = [sum(list) for list in the_data]
    # print(f'column_totals = {column_totals}')

    len_cell = len(the_data[0])
    midpoint = int(len_cell/2)
    
    cell = [[list[:midpoint],list[midpoint:len_cell]] for list in the_data]
    # print(f'cell = {cell}')
    # print(len(cell))
    
    # print([list for list in cell])
    # print([list[n] for n in range(midpoint) for list in cell])

    cell_items = [list[n] for n in range(midpoint) for list in cell]
    print(f'cell_items = {cell_items}')

    n = len(cell_items[0])
    print(f'n = {n}')

    sum_cell_items = [sum(item) for item in cell_items]
    print(f'sum_cell_items = {sum_cell_items}')

    T_sq_cell_over_n = round(sum([num**2/n for num in sum_cell_items]),3)
    print(f'T_sq_cell_over_n = {T_sq_cell_over_n}')

    SS_between = round(T_sq_cell_over_n - G_sq_over_N,3)
    print(f'SS_between = {SS_between}')

    SS_within = round(sum_X_sq - T_sq_cell_over_n,3)
    print(f'SS_within = {SS_within}')

    # sum_cell_items = [16, 14, 18, 20, 34, 42]
    midway = int(len(sum_cell_items)/2)
    list1 = sum_cell_items[:midway]
    list2 = sum_cell_items[midway:]
    sum_T_sq_column = round(sum([((list1[i]+list2[i])**2)/rn for i in range(len(list1))]),3)
    print(f'sum_T_sq_column = {sum_T_sq_column}')

    T_sq_row_cn = round((sum(list1)**2)/cn + (sum(list2)**2)/cn,3)
    print(f'T_sq_row_cn = {T_sq_row_cn}')

    SS_column = round(sum_T_sq_column - G_sq_over_N,3)
    print(f'SS_column = {SS_column}')

    SS_row = round(T_sq_row_cn - G_sq_over_N,3)
    print(f'SS_row = {SS_row}')

    SS_interaction = round(SS_between - (SS_column + SS_row),3)
    print(f'SS_interaction = {SS_interaction}')

    df_total = N - 1
    df_column = c - 1
    df_row = r - 1
    df_interaction = (c-1)*(df_row)
    df_within = N - (c*r)
    print(f'df_total = {df_total}, df_column = {df_column}, df_row = {df_row}, df_interaction = {df_interaction}, df_within = {df_within}')

    MS_column = round(SS_column/df_column,3)
    MS_row = round(SS_row/df_row,3)
    MS_interaction = round(SS_interaction/df_interaction,3)
    MS_within = round(SS_within/df_within,3)
    print(f'MS_column = {MS_column}, MS_row = {MS_row}, MS_interaction = {MS_interaction}, MS_within = {MS_within}')
    
    F_column = round(MS_column/MS_within,3)
    F_row = round(MS_row/MS_within,3)
    F_interaction = round(MS_interaction/MS_within,3)
    print(f'F_column = {F_column}, F_row = {F_row}, F_interaction = {F_interaction}')

    eta_sq_column = round(SS_column/(SS_column+SS_within),3)
    eta_sq_row = round(SS_row/(SS_row+SS_within),3)
    eta_sq_interaction = round(SS_interaction/(SS_interaction+SS_within),3)
    print(f'eta_sq_column = {eta_sq_column}, eta_sq_row = {eta_sq_row}, eta_sq_interaction = {eta_sq_interaction}')

    p_value_column = round(1 - f.cdf(F_column, df_column, df_within),7)
    print(p_value_column<0.05)
    p_value_row = round(1 - f.cdf(F_row, df_row, df_within),7)
    print(p_value_row<0.01)
    p_value_interaction = round(1 - f.cdf(F_interaction, df_interaction, df_within),7)
    print(p_value_interaction<0.05)

In [32]:
X_0 = [8,8,9,11]
X_2 = [8,6,15,19]
X_4 = [10,8,24,18]
list = [X_0,X_2,X_4]
sig = 0.05
row = 2
two_factor_ANOVA(list,sig,row)

joined = [ 8  8  9 11  8  6 15 19 10  8 24 18]
rn = 4
N = 12
G = 144
c = 3
cn = 6
X_sq = [64, 64, 81, 121, 64, 36, 225, 361, 100, 64, 576, 324]
sum_X_sq = 2080
G_sq_over_N = 1728.0
SS_total = 352.0
cell_items = [[8, 8], [8, 6], [10, 8], [9, 11], [15, 19], [24, 18]]
n = 2
sum_cell_items = [16, 14, 18, 20, 34, 42]
T_sq_cell_over_n = 2048.0
SS_between = 320.0
SS_within = 32.0
sum_T_sq_column = 1800.0
T_sq_row_cn = 1920.0
SS_column = 72.0
SS_row = 192.0
SS_interaction = 56.0
df_total = 11, df_column = 2, df_row = 1, df_interaction = 2, df_within = 6
MS_column = 36.0, MS_row = 192.0, MS_interaction = 28.0, MS_within = 5.333
F_column = 6.75, F_row = 36.002, F_interaction = 5.25
eta_sq_column = 0.692, eta_sq_row = 0.857, eta_sq_interaction = 0.636
True
True
True


#### Mean reaction times increase with crowd size [F(2, 6) = 6.75, MSE = 5.33, p < .05, $\eta^2_{p}$ = .69]; they are larger for nondangerous than for dangerous conditions [F((1, 6) = 36.02, p < .01, $\eta^2_{p}$ = .86]; but these findings must be qualifiedbecause of the significant interaction [F(2, 6) = 5.25, p < .05, $\eta^2_{p}$ = .64]

In [31]:
X0 = [0,1,0,0]
X1 = [1,0,1,1]
X2 = [3,5,6,2]
X3 = [6,9,6,10]
tv = [X0,X1,X2,X3]
sig = 0.05
row = 2
two_factor_ANOVA(tv,sig,row)

joined = [ 0  1  0  0  1  0  1  1  3  5  6  2  6  9  6 10]
rn = 4
N = 16
G = 51
c = 4
cn = 8
X_sq = [0, 1, 0, 0, 1, 0, 1, 1, 9, 25, 36, 4, 36, 81, 36, 100]
sum_X_sq = 331
G_sq_over_N = 162.5625
SS_total = 168.438
cell_items = [[0, 1], [1, 0], [3, 5], [6, 9], [0, 0], [1, 1], [6, 2], [6, 10]]
n = 2
sum_cell_items = [1, 1, 8, 15, 0, 2, 8, 16]
T_sq_cell_over_n = 307.5
SS_between = 144.938
SS_within = 23.5
sum_T_sq_column = 306.75
T_sq_row_cn = 162.625
SS_column = 144.188
SS_row = 0.062
SS_interaction = 0.688
df_total = 15, df_column = 3, df_row = 1, df_interaction = 3, df_within = 8
MS_column = 48.063, MS_row = 0.062, MS_interaction = 0.229, MS_within = 2.938
F_column = 16.359, F_row = 0.021, F_interaction = 0.078
eta_sq_column = 0.86, eta_sq_row = 0.003, eta_sq_interaction = 0.028
True
False
False


In [5]:
def HSD_two_factor_anova(MS_within,df_within,n,k,alpha):
    import statsmodels.stats.libqsturng as qsturng
    q = round(qsturng.qsturng(1 - alpha, k, df_within),3)
    print(f'q = {q}')
    Tukeys_HSD_two_factor_anova = round(q*(MS_within/n)**0.5,3)
    print(f'Tukeys_HSD_two_factor_anova = {Tukeys_HSD_two_factor_anova}')

MS_within = 2.938
df_within = 8
n = 4 #can be row numbers(rn) or column numbers(cn)
k = 4
alpha = 0.05
HSD_two_factor_anova(MS_within,df_within,n,k,alpha)

q = 4.529
Tukeys_HSD_two_factor_anova = 3.881


In [7]:
plain = [7.2,8.9]
vegetarian = [5.7,4.8]
salami = [4.8,8.4]
everything = [6.1,1.3]
data = [plain,vegetarian,salami,everything]
sig = 0.5
row = 2

two_factor_ANOVA(data,sig,row)
# print('==========')
# Anova_pvalue_HSD(data,sig)
# print('==========')
# Repeated_ANOVA(data,sig)

joined = [7.2 8.9 5.7 4.8 4.8 8.4 6.1 1.3]
rn = 2
N = 8
G = 47.2
c = 4
cn = 8
X_sq = [51.84, 79.21, 32.49, 23.04, 23.04, 70.56, 37.21, 1.69]
sum_X_sq = 319.08
G_sq_over_N = 278.48
SS_total = 40.6
cell_items = [[7.2], [5.7], [4.8], [6.1]]
n = 1
sum_cell_items = [7.2, 5.7, 4.8, 6.1]
T_sq_cell_over_n = 144.58
SS_between = -133.9
SS_within = 174.5
sum_T_sq_column = 141.62
T_sq_row_cn = 35.652
SS_column = -136.86
SS_row = -242.828
SS_interaction = 245.788
df_total = 7, df_column = 3, df_row = 1, df_interaction = 3, df_within = 0
MS_column = -45.62, MS_row = -242.828, MS_interaction = 81.929, MS_within = inf
F_column = -0.0, F_row = -0.0, F_interaction = 0.0
eta_sq_column = -3.636, eta_sq_row = 3.554, eta_sq_interaction = 0.585


/tmp/ipykernel_29359/2375349913.py:95: RuntimeWarning: divide by zero encountered in scalar divide
  MS_within = round(SS_within/df_within,3)


In [8]:
X0 = [0,1,0,0]
X1 = [1,0,1,1]
X2 = [3,5,6,2]
X3 = [6,9,6,10]
tv = [X0,X1,X2,X3]
sig = 0.05
row = 2
two_factor_ANOVA(tv,sig,row)

joined = [ 0  1  0  0  1  0  1  1  3  5  6  2  6  9  6 10]
rn = 4
N = 16
G = 51
c = 4
cn = 8
X_sq = [0, 1, 0, 0, 1, 0, 1, 1, 9, 25, 36, 4, 36, 81, 36, 100]
sum_X_sq = 331
G_sq_over_N = 162.5625
SS_total = 168.438
cell_items = [[0, 1], [1, 0], [3, 5], [6, 9], [0, 0], [1, 1], [6, 2], [6, 10]]
n = 2
sum_cell_items = [1, 1, 8, 15, 0, 2, 8, 16]
T_sq_cell_over_n = 307.5
SS_between = 144.938
SS_within = 23.5
sum_T_sq_column = 306.75
T_sq_row_cn = 162.625
SS_column = 144.188
SS_row = 0.062
SS_interaction = 0.688
df_total = 15, df_column = 3, df_row = 1, df_interaction = 3, df_within = 8
MS_column = 48.063, MS_row = 0.062, MS_interaction = 0.229, MS_within = 2.938
F_column = 16.359, F_row = 0.021, F_interaction = 0.078
eta_sq_column = 0.86, eta_sq_row = 0.003, eta_sq_interaction = 0.028
